### PR TITLE
feat: Alerts! allowing users to set SQL-based email alerts with screenshots

### PR DESCRIPTION
**Pull request by mistercrunch**
This is a collaboration with @robdiciuccio 

## Screenshots!

Screen Shot 2020-05-29 at 12 07 25 AM
Screen Shot 2020-05-29 at 12 12 22 AM
Screen Shot 2020-05-29 at 12 07 41 AM
Screen Shot 2020-05-29 at 12 07 47 AM

## Components
- 2 new models: `Alert` and `AlertLog`
- A scheduler loop, triggered by Celery Beat
- 2 new endpoints 
    - Cache a screenshot `/api/v1/{chart_pk}/cache_screenshot/?q=`
    - View a screenshot `/api/v1/{chart_pk}/screenshot/{SHA}`
- Improvements to `BaseScreenshot`

## Future / TODO
- Hook up `grace_period` parameter
- Hook up `retention` parameter
- Cleanup
- Slack!

### CODE REVIEW

**Review by kalimuthu123 — APPROVED** [review]
good work is it support conditional alerting

**Comment by rumbin** [comment]
Great feature! I'm looking forward to having it in an official release :-)

**Review by kalimuthu123 — APPROVED** [review]
good work

File: superset/app.py
Comment (willbarrett) [inline superset/app.py:386]: Should this be a config flag or a feature flag?

File: superset/charts/schemas.py
Comment (willbarrett) [inline superset/charts/schemas.py:30]: nit: typo - should be `width_height_schema`

File: superset/tasks/schedules.py
Comment (willbarrett) [inline superset/tasks/schedules.py:94]: Let's not write email addresses to logs. This is protected information in most organizations.

**Review by willbarrett — COMMENTED** [review]
Nice start!

File: superset/charts/api.py
Comment (bkyryliuk) [inline superset/charts/api.py:594]: maybe return 404 ?

File: superset/charts/api.py
Comment (bkyryliuk) [inline superset/charts/api.py:544]: s/screenshot/cached_screenshot?

File: superset/charts/api.py
Comment (bkyryliuk) [inline superset/charts/api.py:467]: s/cache_screenshot/schedule_screenshot

File: superset/charts/api.py
Comment (bkyryliuk) [inline superset/charts/api.py:468]: s/Get Chart screenshot/Schedule screenshot computation in celery

File: superset/charts/api.py
Comment (bkyryliuk) [inline superset/charts/api.py:640]: maybe add chart id to the logging message

File: superset/cli.py
Comment (bkyryliuk) [inline superset/cli.py:599]: delete me :) 
it would be useful to have a dev cli for things like it

File: superset/migrations/versions/2f1d15e8a6af_add_alerts.py
Comment (bkyryliuk) [inline superset/migrations/versions/2f1d15e8a6af_add_alerts.py:69]: s/alert_owner/alert_owners
if I get it right alert can have many owners

File: superset/models/alerts.py
Comment (bkyryliuk) [inline superset/models/alerts.py:47]: would be nice to have AuditMixinNullable

File: superset/tasks/schedules.py
Comment (bkyryliuk) [reply superset/tasks/schedules.py:94]: maybe do it debug logging? so it is configurable if needed ?

File: superset/tasks/schedules.py
Comment (bkyryliuk) [inline superset/tasks/schedules.py:504]: nit: would be nicer to raise here

File: superset/tasks/schedules.py
Comment (bkyryliuk) [inline superset/tasks/schedules.py:647]: probably just delete it, I assume this function should not be called if it's not enabled

File: superset/utils/screenshots.py
Comment (bkyryliuk) [inline superset/utils/screenshots.py:264]: would be nice to have a timeout here and time to live here
e.g. for hourly alerts ttl should be 1 hour tops

**Review by bkyryliuk — APPROVED** [review]
LG%nits

It is a good foundation to continue iterating on the feature.
We can take it from here and start polishing and adding more tests.
It would be great to merge the PR.

**Comment by kkalyan** [comment]
Nice! SQL check is powerful and compact. 

Does this replace current Email notification feature for dashboards and slices?

**Comment by bkyryliuk** [comment]
@mistercrunch, @robdiciuccio how would you prefer to take it forward? 
This PR touches many files and it would be hard to keep it up to date if this one doesn't get merged.
From my prospective we would love to see it merged and contribute towards alerting in superset, e.g. we would start with adding tests & making it more robust.

**Comment by robdiciuccio** [comment]
Thanks for the ping @bkyryliuk. I'll get this cleaned up and merged.

File: superset/charts/api.py
Comment (robdiciuccio) [inline superset/charts/api.py:611]: We'll need to add the appropriate permissions check here (and to all screenshot endpoints) before this can merge.

File: superset/tasks/schedules.py
Comment (robdiciuccio) [reply superset/tasks/schedules.py:613]: We likely don't want celery to retry the failed jobs in this case.

File: superset/charts/api.py
Comment (robdiciuccio) [reply superset/charts/api.py:611]: Because these endpoints are behind the `THUMBNAILS` feature flag, which defaults to `False`, should be OK to merge now in order to unblock.

**Comment by robdiciuccio** [comment]
@bkyryliuk @willbarrett this looks good to me as a first pass, with the understanding that @bkyryliuk's team will be iterating on it. If it LGTY, feel free to merge.

**Review by willbarrett — APPROVED** [review]

**Comment by bkyryliuk** [comment]
yep, big thanks @robdiciuccio and @willbarrett !

**Comment by mistercrunch** [comment]
tenor (1)

**Comment by heyuan0403** [comment]
Hi, thanks for adding this great feature ! Does superset have a plan for official release of this ?

**Comment by camilo-uptime** [comment]
Hi!, thanks for this feature!, I've been trying to test it with no success I enabled the feature on the config, and created a dummy alert, spawned a celery worker and beat but nothing happens, did I miss something?, I have chart and dashboard reports working properly :(.

**Comment by mistercrunch** [comment]
There's a CLI command `superset alert` that runs a scheduler loop and can help test the feature.

**Comment by brachipa** [comment]
Same for me, it doesn't work, also by trying `superset alert`
I see this exception in logs 
```stacktrace
[2020-10-01 20:17:45,040: INFO/ForkPoolWorker-1] Processing alert ID: 2
[2020-10-01 20:17:45,042: DEBUG/ForkPoolWorker-1] Parsing with sqlparse statement: select 1,0,5,55
[2020-10-01 20:17:45,044: INFO/ForkPoolWorker-1] Evaluating SQL for alert <2:brachi>
[2020-10-01 20:17:45,045: DEBUG/ForkPoolWorker-1] Database.get_sqla_engine(). Masked URL: postgresql://superset:XXXXXXXXXX@db:5432/superset
[2020-10-01 20:17:45,054: INFO/ForkPoolWorker-1] Triggering alert: <2:brachi>
[2020-10-01 20:17:45,057: ERROR/ForkPoolWorker-1] Task alerts.run_query[72f7c16b-ba6f-4836-916d-222fb3146641] raised unexpected: OSError(99, 'Cannot assign requested address')
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/celery/app/trace.py", line 412, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/app/superset/app.py", line 115, in __call__
    return task_base.__call__(self, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/celery/app/trace.py", line 704, in __protected_call__
    return self.run(*args, **kwargs)
  File "/app/superset/tasks/schedules.py", line 554, in schedule_alert_query
    if run_alert_query(schedule, dbsession):
  File "/app/superset/tasks/schedules.py", line 648, in run_alert_query
    deliver_alert(alert)
  File "/app/superset/tasks/schedules.py", line 607, in deliver_alert
    _deliver_email(alert.recipients, deliver_as_group, subject, body, data, images)
  File "/app/superset/tasks/schedules.py", line 132, in _deliver_email
    dryrun=config["SCHEDULED_EMAIL_DEBUG_MODE"],
  File "/app/superset/utils/core.py", line 787, in send_email_smtp
    send_mime_email(smtp_mail_from, recipients, msg, config, dryrun=dryrun)
  File "/app/superset/utils/core.py", line 808, in send_mime_email
    else smtplib.SMTP(smtp_host, smtp_port)
  File "/usr/local/lib/python3.6/smtplib.py", line 251, in __init__
    (code, msg) = self.connect(host, port)
  File "/usr/local/lib/python3.6/smtplib.py", line 336, in connect
    self.sock = self._get_socket(host, port, self.timeout)
  File "/usr/local/lib/python3.6/smtplib.py", line 307, in _get_socket
    self.source_address)
  File "/usr/local/lib/python3.6/socket.py", line 724, in create_connection
    raise err
  File "/usr/local/lib/python3.6/socket.py", line 713, in create_connection
    sock.connect(sa)
OSError: [Errno 99] Cannot assign requested address
```

**Comment by mistercrunch** [comment]
Email configuration doesn't seem to work here.Looks like you need to open your SMTP port

**Pull request merged**
